### PR TITLE
feat(tooling): explicit encoding="utf-8" at every text-mode file call + gate (#645)

### LINKED ISSUE #645: Audit text-mode open()/read_text()/write_text() for explicit encoding='utf-8' (#641 follow-up)
**Origin:** PR for #641 (cp1252-on-Windows fix) chose Option B (set `PYTHONUTF8=1` in CI) as the primary fix because Option A (audit every `open()` / `read_text()` / `write_text()` call site for explicit `encoding='utf-8'`) was too large for the same PR's scope.  This issue tracks Option A as the durable follow-up.

## The pattern

Python on Windows uses `locale.getpreferredencoding()` for text-mode `open()` / `read_text()` / `write_text()` calls without an explicit `encoding=` kwarg.  In en-US Windows that's cp1252; in other locales it's whatever the system default is.  cp1252 can't represent most of Unicode; UTF-8 can.

Vera's source / docs / fixtures contain Unicode characters in:

- Test fixtures with `→` (right arrow) and `—` (em-dash) in error messages, contract violations, trap diagnostics
- HTML examples in `docs/index.html`
- Regenerated `llms.txt` / `SKILL.md` outputs that contain prose from CHANGELOG / spec
- Any `.vera` file that contains non-ASCII (string literals, comments)

PR for #641 covered CI via `PYTHONUTF8=1` and added explicit `encoding='utf-8'` to the load-bearing `vera/parser.py` grammar load.  Per-call-site coverage for the rest of the codebase is the durable fix — locally users on Windows without `PYTHONUTF8=1` still hit the bug on individual files.

## Audit surface (rough, from the PR-for-#641 grep)

~30+ sites across `vera/`, `scripts/`, `tests/`:

```
vera/parser.py — already done (PR for #641)
vera/browser/emit.py — html_path.write_text(...)
scripts/check_conformance.py — MANIFEST_PATH.read_text()
scripts/check_version_sync.py — pyproject.read_text(), init.read_text(), index_html.read_text(), readme.read_text(), lock.read_text()
scripts/check_doc_counts.py — manifest.json, .pre-commit-config.yaml, ci.yml, TESTING.md, CONTRIBUTING.md, CLAUDE.md, README.md, SKILL.md, AGENTS.md, FAQ.md, docs/index.html, ROADMAP.md
scripts/check_limitations_sync.py — KNOWN_ISSUES.md, vera/README.md
tests/test_codegen_monomorphize.py — path.read_text() (×2)
tests/test_codegen_closures.py — path.read_text() (×2)
tests/test_tester_coverage.py — p.write_text(source)
```

## Recommended approach

A single audit-and-replace pass with a small ruff or grep-based check to enforce going forward:

1. Mechanical: grep + add `encoding='utf-8'` to every text-mode site.
2. Add a pre-commit check (or a CI step) that fails on `open(...)`, `read_text()`, `write_text()` without an explicit encoding.  Use a small `scripts/check_explicit_encoding.py` that grep-greps the codebase.

## Out of scope

- Binary-mode opens (`'rb'`, `'wb'`) — encoding is irrelevant.
- `urlopen` / `subprocess` / network — no text encoding.
- Third-party calls (e.g. `lark.Lark(...)`) that take an already-decoded string — encoding handled at the read site.

## Acceptance criteria

- Every text-mode `open()` / `read_text()` / `write_text()` in `vera/`, `scripts/`, `tests/` has explicit `encoding='utf-8'`.
- A pre-commit check enforces the convention.
- The `PYTHONUTF8: 1` line in `.github/workflows/ci.yml`'s test job env can be removed (verified by a clean CI run without it).

## Pairs with

- **#641** — landed in v0.0.143 with the Option B (CI-side) fix.

### LINKED ISSUE #832: Route CLI stdout/stderr + subprocess text I/O through UTF-8 so PYTHONUTF8 can be dropped (AC3 of #645)
**Split out from [#645](https://github.com/aallan/vera/issues/645) (AC3).**  #645 landed the file-I/O half — every text-mode `open()` / `read_text()` / `write_text()` now passes an explicit `encoding="utf-8"`, enforced by `scripts/check_explicit_encoding.py` — and attempted AC3 (remove the `PYTHONUTF8=1` CI backstop).  **AC3 reddened the Windows matrix** and was reverted.

## Why the file-I/O audit wasn't enough

`PYTHONUTF8` (PEP 540) also governs **stdio and subprocess** text encoding, not just `open()`.  On a cp1252 Windows runner without it:

- A Vera program that does `IO.print("→")` writes `→` (U+2192) to the child process's `sys.stdout`, which defaults to cp1252 → `UnicodeEncodeError: 'charmap' codec can't encode character '\u2192'`.
- Tests that capture such output via `subprocess.run(..., text=True)` decode the child's bytes with the parent's cp1252 → `UnicodeDecodeError: 'utf-8' codec can't decode byte 0x97` (0x97 = cp1252 em-dash).

Evidence: the [#645 PR](https://github.com/aallan/vera/pull/831) CI run 28515315521 — Windows 3.11/3.12/3.13 failed with exactly these errors once `PYTHONUTF8` was removed.

## Scope

- Reconfigure the `vera` CLI's `sys.stdout` / `sys.stderr` to UTF-8 (e.g. `.reconfigure(encoding="utf-8")`) so a Vera program's non-ASCII output is UTF-8 regardless of the host locale.
- Pass `encoding="utf-8"` (or decode bytes explicitly) at every `subprocess` text-capture site in `tests/` that round-trips program output.
- Then remove `PYTHONUTF8: 1` from `.github/workflows/ci.yml` and verify a clean Windows matrix (AC3's own acceptance test).

## Out of scope

- The file-I/O audit + gate (done in #645).

### PR DESCRIPTION
**Pull request by aallan**
## What & why

Python's text-mode `open()` / `Path.read_text()` / `Path.write_text()` fall back to `locale.getpreferredencoding()` (cp1252 on en-US Windows) when no `encoding=` is given — so a Vera source / doc / fixture containing `→`, `—`, or any non-ASCII byte failed to read/write on a locale-default Windows shell. The same locale fallback governs `subprocess(..., text=True)` captures and the CLI child process's own `sys.stdout`. [#641](https://github.com/aallan/vera/issues/641) papered over CI with a `PYTHONUTF8=1` backstop; #645 is the durable per-site fix that lets the backstop go.

## Changes

1. **Explicit `encoding="utf-8"` at every text-mode file site (AC1)** — new `scripts/check_explicit_encoding.py` AST-audits every text-mode `open()` / `read_text()` / `write_text()` under `vera/`, `scripts/`, `tests/` (binary skipped; `.open()` out of scope; `# encoding-exempt: <reason>` opt-out; the `encoding=` value must be a `"utf-8"` literal). It found **160 bare file sites across 16 files** — ~5× the issue's estimate — all now fixed.
2. **Subprocess text captures + CLI stdio (AC3, done here)** — the gate was extended to also flag `subprocess.run` / `Popen` / `check_output(..., text=True | universal_newlines=True)` without an explicit `encoding="utf-8"` (bytes mode is skipped — `bytes.decode()` defaults to UTF-8, so no locale hazard). **97 subprocess sites across 18 files** fixed. `vera/cli.py` `main()` now reconfigures `sys.stdout` / `sys.stderr` to UTF-8 at startup (guarded via `getattr`), so a Vera program that prints `→` / `—` is UTF-8 regardless of host locale.
3. **A gate enforces all of it (AC2)** — wired into pre-commit (`explicit-encoding`) + the CI `lint` job. `tests/test_check_explicit_encoding.py` pins the checker logic (file I/O **and** subprocess) **and** asserts the repo is clean.
4. **`PYTHONUTF8=1` removed from the CI matrix (AC3)** — with text file I/O, subprocess captures, and CLI stdio all explicitly UTF-8, the backstop is no longer load-bearing. The Windows matrix (no `PYTHONUTF8`) is the acceptance test.

All behaviour-preserving: the 160 + 97 additions don't change behaviour on a UTF-8 host; full suite is green.

## History (rider)

Trimmed the `v0.0.188` / `v0.0.189` HISTORY rows to single sentences.

## Exit checklist

- [x] `check_explicit_encoding.py` — repo clean (file I/O + subprocess); gate wired (pre-commit + CI `lint`)
- [x] full `pytest` — 5,492 passed, 22 skipped (behaviour-preserving); 37 checker tests
- [x] `mypy` · `ruff` · doc-counts (5,540 / 52 / 31 hooks / 8 CI jobs) · version-sync (v0.0.190 ×6) · site-assets — green
- [x] CHANGELOG `[0.0.190]` · ROADMAP #645 + #832 → removed · HISTORY v0.0.190 row
- [ ] CI matrix green **with `PYTHONUTF8` removed** — the AC3 acceptance test (running)

Closes #645
Closes #832

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Enforced explicit UTF-8 for all text-mode file reads/writes, with an inline opt-out marker for exceptional cases.
  * Added repository checks for subprocess text capture to ensure consistent UTF-8 output, and ensured the CLI emits UTF-8 on standard streams.

* **Bug Fixes**
  * Improved cross-platform determinism by removing reliance on platform-default text encodings.
  * Removed the previous CI UTF-8 backstop now that enforcement is in place.

* **Tests**
  * Added coverage for the explicit-encoding enforcement and updated suites to use UTF-8 consistently.

* **Documentation**
  * Updated README, contributing/testing guidance, changelog, roadmap, and history for release **0.0.190**.

### CODE REVIEW

**Comment by aallan** [comment]
## pr-review-toolkit pass — clean, AC3 empirically de-risked

A `code-reviewer` pass focused on the new logic (skipping the 160 mechanical edits) found **no actionable issues**. The checker's AST logic (text-mode flag / binary skip / `read_bytes` skip / `.open()` out-of-scope / non-literal-mode / tokenize-based `# encoding-exempt` opt-out incl. the string-vs-comment distinction) is correct, and the tests pin it plus the repo-clean assertion.

**The riskiest change — AC3 (`PYTHONUTF8` removal) — was verified by execution, not inference.** `PYTHONUTF8` also governs stdio/subprocess encoding, so the review probed every CLI output path with a forced-`cp1252` stderr:
- The `vera` CLI emits **pure ASCII** on every human-readable path (`errors`/`builtins`/`effects`/`check`/`verify`/`--explain-slots`). The only non-ASCII byte reachable in output is one en-dash `–` (U+2013, in the `E001–E702` USAGE string) — and U+2013 **is** representable in cp1252, so it does not raise. Re-running `cli.main()` with `sys.stderr` forced to `cp1252`/strict on the no-args / unknown-command paths kept them green.
- `--json` output is locale-safe: every `json.dumps` uses the default `ensure_ascii=True` (non-ASCII → `\uXXXX`).
- The two subprocess tests that round-trip `é` (`test_cli.py:1932`, `:1960`) already pass `encoding="utf-8"`; `test_lsp.py` uses byte pipes with hardcoded utf-8; `test_runtime_traps.py:2557`'s child is ASCII-only.

So the audit's file-I/O scope is sufficient — no stdio/subprocess Unicode path depends on `PYTHONUTF8`. The Windows matrix on this PR is the final confirmation.

Minor sub-threshold notes (no action): positional-`encoding` arg would be a theoretical false-positive (no such call in-repo); aliased `f = open` / `io.open` aren't detected (inherent AST-gate limit, documented).

File: HISTORY.md [hunk 1]
```diff
@@ -412,8 +412,9 @@ A sustained audit of every place where `vera verify` could prove a postcondition
...
-| v0.0.189 | 1 Jul | The UTF-8 "safe decode" invariant is centralised into one `safe_utf8_decode` helper, reached only through a shared `_slice_and_decode` helper that the three WASM-memory string readers (which the host imports delegate to) route through, replacing six brittle source-grep tests with a unit test plus three wire-real end-to-end tests ([#592](https://github.com/aallan/vera/issues/592)). |
+| v0.0.188 | 30 Jun | Every diagnostic is gated for complete, spec-accurate `rationale` / `fix` / `spec_ref` metadata ([#682](https://github.com/aallan/vera/issues/682)). |
+| v0.0.189 | 1 Jul | The UTF-8 "safe decode" invariant is centralised behind one `safe_utf8_decode` helper, replacing six brittle source-grep tests with behavioural coverage ([#592](https://github.com/aallan/vera/issues/592)). |
+| v0.0.190 | 1 Jul | Every text-mode file call now passes an explicit `encoding="utf-8"`, enforced by a pre-commit/CI gate ([#645](https://github.com/aallan/vera/issues/645)). |
```
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md]: @coderabbitai Fixed in 12c0016 — narrowed the v0.0.190 row from "Every text-mode file call" to the three audited forms: "Text-mode `open` / `read_text` / `write_text` calls now pass an explicit `encoding="utf-8"`, enforced by a pre-commit/CI gate". (Also tightened the CHANGELOG headline the same way.)

File: scripts/check_explicit_encoding.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            elif text.startswith(OPT_OUT) and (
+                    text[len(OPT_OUT)] == ":" or text[len(OPT_OUT)].isspace()):
+                out[tok.start[0]] = text[len(OPT_OUT):].lstrip(" :").strip()
+    except (tokenize.TokenError, IndentationError):
+        pass
+    return out
+
+
+def _has_encoding(call: ast.Call) -> bool:
+    return any(kw.arg == "encoding" for kw in call.keywords)
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_explicit_encoding.py]: @coderabbitai Good catch — fixed in 12c0016. The gate now requires the `encoding=` value to be a UTF-8 **string literal**: `_is_utf8_literal` accepts `"utf-8"` / `"UTF-8"` / `"utf8"` (case-insensitive, hyphen/underscore-normalised) and rejects a non-literal (`encoding=enc`) or a different codec (`encoding="latin-1"`), across all three call forms. A deliberate non-UTF-8 site uses `# encoding-exempt: <reason>` rather than a bare `encoding=`. Verified the repo stays clean (all 298 real sites are `"utf-8"`); tests updated (+5: case-insensitive spellings pass; non-utf-8 / non-literal are flagged).

**Comment by aallan** [comment]
@coderabbitai Re the outside-diff finding on `TESTING.md` (hook table: 28 rows vs the "29 commit-stage" heading): the heading count is correct — the `check-yaml / check-toml` row bundled **two** hooks, so 28 rows enumerated 29 commit-stage hooks (31 total incl. the 2 push-stage; gate-verified by `check_doc_counts`). To remove the ambiguity, 12c0016 splits that row into separate `check-yaml` and `check-toml` rows, so the row count now matches the hook count.

**Comment by aallan** [comment]
@coderabbitai full review

**Comment by aallan** [comment]
@coderabbitai full review

Pushed `c12629a`: AC3 is now done in-PR (previously split to #832, folded back). Extended `check_explicit_encoding.py` to also flag `subprocess.run/Popen/check_output(..., text=True)` captures without an `encoding="utf-8"` literal (bytes mode skipped), fixed all 97 flagged subprocess sites, added a startup `sys.stdout/stderr` UTF-8 reconfigure in `vera/cli.py`, and removed `PYTHONUTF8=1` from the CI matrix. The Windows matrix without the backstop is the acceptance test.

File: tests/test_check_explicit_encoding.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+# ---------------------------------------------------------------------------
+# The repository itself is clean (the production assertion).
+# ---------------------------------------------------------------------------
+
+def test_repository_has_no_bare_text_calls(mod: object) -> None:
+    violations = mod.check_paths(mod.iter_scope_files())
+    assert violations == [], (
+        f"{len(violations)} bare text-mode file call(s) lack encoding='utf-8' "
+        "(#645):\n" + "\n".join(
+            f"  {v.file}:{v.line}: {v.reason}" for v in violations))
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_check_explicit_encoding.py:254]: @coderabbitai Fixed in 0d9d862. Added `test_iter_scope_files_reaches_every_scope_root` (asserts every `SCOPE_DIRS` root is reached and a known real file under each is discovered) and `test_in_scope_accepts_scope_roots_and_rejects_outsiders`. Mutation-validated: dropping `tests` from `SCOPE_DIRS`, `iter_scope_files`→`[]`, and `_in_scope`→always-True each turn the new tests RED while the old clean-repo assertion stays GREEN — exactly the gap you flagged.

File: HISTORY.md [hunk 4]
```diff
@@ -412,8 +412,9 @@ A sustained audit of every place where `vera verify` could prove a postcondition
...
-| v0.0.189 | 1 Jul | The UTF-8 "safe decode" invariant is centralised into one `safe_utf8_decode` helper, reached only through a shared `_slice_and_decode` helper that the three WASM-memory string readers (which the host imports delegate to) route through, replacing six brittle source-grep tests with a unit test plus three wire-real end-to-end tests ([#592](https://github.com/aallan/vera/issues/592)). |
+| v0.0.188 | 30 Jun | Every diagnostic is gated for complete, spec-accurate `rationale` / `fix` / `spec_ref` metadata ([#682](https://github.com/aallan/vera/issues/682)). |
+| v0.0.189 | 1 Jul | The UTF-8 "safe decode" invariant is centralised behind one `safe_utf8_decode` helper, replacing six brittle source-grep tests with behavioural coverage ([#592](https://github.com/aallan/vera/issues/592)). |
+| v0.0.190 | 1 Jul | Text I/O (files and `subprocess` captures) is UTF-8 regardless of host locale, gated in pre-commit/CI, letting the `PYTHONUTF8` backstop be removed ([#645](https://github.com/aallan/vera/issues/645)). |
```
[the remark this answers is not in the document]
Comment (aallan) [reply HISTORY.md]: @coderabbitai Fixed in 5ab25a7. The v0.0.188 HISTORY row now reads "spec-accurate `rationale` and `spec_ref` metadata, plus a `fix` for errors" — matching `check_diagnostic_fields.py`, which exempts warnings from the `fix` requirement.

File: vera/cli.py [hunk 5]
```diff
@@ -1249,6 +1249,16 @@ def cmd_errors(as_json: bool = False) -> int:
...
+    # Emit program output and diagnostics as UTF-8 regardless of the host
+    # locale: a Vera program printing `→` / `—` (or any non-ASCII) must not hit
+    # cp1252 on a locale-default Windows shell (#645).  Guarded via getattr — a
+    # wrapped or replaced stream (e.g. pytest's capture) may not expose
+    # reconfigure(), and encoding to UTF-8 never raises (it covers all Unicode).
+    for _stream in (sys.stdout, sys.stderr):
+        _reconfigure = getattr(_stream, "reconfigure", None)
+        if _reconfigure is not None:
+            _reconfigure(encoding="utf-8")
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/cli.py:1262]: @coderabbitai Already fixed in 409f847 (this thread is anchored to the earlier c12629a). The loop is now `for _stream in (sys.stdin, sys.stdout, sys.stderr)`, so `IO.read_char` / `read_line` / stdin-backed loading decode UTF-8 regardless of host locale — this closed the 19th Windows failure (`test_run_io_read_char_utf8`).

File: scripts/check_explicit_encoding.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    """True iff a subprocess call decodes its captured output as text via the
+    locale codec — i.e. ``text=True`` or ``universal_newlines=True`` is set.
+    Without either it returns ``bytes`` (and ``bytes.decode()`` defaults to
+    UTF-8, so there is no locale hazard)."""
+    for kw in call.keywords:
+        if kw.arg in ("text", "universal_newlines"):
+            v = kw.value
+            if isinstance(v, ast.Constant) and v.value is True:
+                return True
+    return False
```
[the remark this answers is not in the document]
Comment (aallan) [reply scripts/check_explicit_encoding.py:165]: @coderabbitai Fixed in 5ab25a7. `_subprocess_is_text` now also treats `encoding=` / `errors=` as text-mode triggers (either forces text mode per the subprocess docs), so `subprocess.run(cmd, encoding="latin-1")` is flagged. Added tests: `encoding="latin-1"`-alone and `errors="replace"`-alone are flagged, `encoding="utf-8"`-alone passes. 0 existing repo sites are newly flagged (all real captures already pass `text=True, encoding="utf-8"`).

File: TESTING.md [hunk 7]
```diff
@@ -649,6 +651,7 @@ Every push is checked by 30 configured hooks across two stages: 28 are configure
 | `check_doc_counts.py` | Counts in docs match live codebase |
 | `check_walker_coverage.py` | Every walker function covers every `Expr` subclass via dispatch or checklist comment (#597) |
 | `check_diagnostic_fields.py` | Every diagnostic in `vera/` carries rationale + spec_ref, and errors also a `fix` (warnings exempt); every `error_code` is registered in `ERROR_CODES` (#828); otherwise a `# diag-fields-exempt` reason (#682) |
+| `explicit-encoding` | Every text-mode `open()` / `read_text()` / `write_text()` passes `encoding="utf-8"` (#645) |
```
[the remark this answers is not in the document]
Comment (aallan) [reply TESTING.md]: @coderabbitai Fixed in d932554. The `explicit-encoding` hook-table row and the "§ file encoding" prose now name `subprocess.run/Popen/check_output` text captures and text-mode `tempfile.NamedTemporaryFile` alongside file I/O, and note the CLI reconfigures stdin as well as stdout/stderr — so the summary matches the gate.

**Comment by aallan** [comment]
@coderabbitai full review

Pushed d932554 (4 commits since the last review): closed the two real AC3 gaps the Windows matrix caught — text-mode `tempfile.NamedTemporaryFile` (29 sites, gate extended to cover it) and `sys.stdin` reconfigure — plus this round's findings (`_subprocess_is_text` now treats `encoding=`/`errors=` as text-mode triggers, precise v0.0.188 HISTORY wording, and the TESTING hook-table synced to the widened gate). The Windows matrix without `PYTHONUTF8` is the acceptance test.

**Comment by aallan** [comment]
@coderabbitai Thanks for the thorough verification. On the one non-blocking note (the `encoding=`/`errors=`-alone text-mode edge case): rather than just documenting it, I closed it in 5ab25a7 — `_subprocess_is_text` now returns text-mode when `encoding=` or `errors=` is present (either forces text mode per the subprocess docs), so `subprocess.run(cmd, encoding="latin-1")` is now flagged; +3 unit tests, and I re-confirmed 0 live sites are affected. The two inline threads on this (subprocess helper, TESTING hook-table) are addressed in 5ab25a7 / d932554.
